### ISSUE TITLE
Vibration doesn't work for some controllers

### ISSUE THREAD

**Issue by toby3d** [godotengine/godot#14634]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**  1c0007b commit from master branch, Linux Mint Mate 18.2 x86_64
**Issue description:** (for example) `Input.start_joy_vibration(0, 0.5, 1, 5)` don't do anything. Tested with connected PS, XBox and Steam Constrollers.

*Bugsquad edit (keywords for easier searching): vibrate, gamepad, joystick*

**Comment by vnen** [comment]
Works for me on Windows 10, with an Xbox 360 controller. Might be Linux only.

**Comment by toby3d** [comment]
The controllers are successfully connected, the control works without unnecessary movements. Vibration works on Steam, but not in Godot.

**Comment by akien-mga** [comment]
Tested on Mageia 6 x86_64 with the same commit and my two controllers work fine:
- Cheap CSL Computer PC / PS3 controller, seen in the [joypads demo](https://github.com/godotengine/godot-demo-projects/tree/master/misc/joypads) as "GreenAsia Inc. USB Joystick"
- Steam Controller, emulating Xbox mode via SC Controller

**Comment by brainsick** [comment]
Ubuntu 17.10 x86_64
Microsoft X-Box One Pad
Godot 2.1.4 (Steam) and Godot 3.0 Beta 1
Joypads Demo - Start Vibration
... works for me

**Comment by toby3d** [comment]
Hm... I try check demo project little later, maybe I just use method not right.

**Comment by toby3d** [comment]
Okay, only Steam Controller via SC Controller app it works with vibration. Others - worked and finded by godot app, but vibration doesn't work.

**Comment by akien-mga** [comment]
We have now entered release freeze for Godot 3.0 and want to focus only on release critical issues for that milestone. Therefore, we're moving this issue to the 3.1 milestone, though a fix may be made available for a 3.0.x maintenance release after it has been tested in the master branch during 3.1 development. If you consider that this issue is critical enough to warrant blocking the 3.0 release until fixed, please comment so that we can assess it more in-depth.

**Comment by akien-mga** [comment]
We need more information on what *exact* controller models and setup do not work, as so far nobody could reproduce those issues.

**Comment by alex-evelyn** [comment]
I am here to bring an exact setup for my controller. I am using MacOS Mojave 10.14.6 with a Dualshock PS4 Controller I bought at Walmart.

The controller vibrates when told to by Steam (Identify Controller Feature). The model number of the controller is CUH-2CT2U. It is an official Sony controller.

Using the System Information app on my Mac, I found this info (Bluetooth).

```
PS4 Controller:
  Address:	90-89-5F-B8-64-E5
  Major Type:	Peripheral
  Minor Type:	Peripheral
  Services:	Wireless Controller
  Paired:	Yes
  Configured:	Yes
  Connected:	Yes
  Manufacturer:	Unknown (0x7, 0x00)
  Firmware Version:	0x0100
  Vendor ID:	0x054C
  Product ID:	0x09CC
  Class of Device:	0x05 0x02 0x2508
  AFH:	On
  AFH Map:	F0F719000370807D
  RSSI:	-42
  Role:	Master
  Connection Mode:	Active Mode
  Interval:	0 ms
  Host Connectable:	No
  EDR Supported:	Yes
  eSCO Supported:	Yes
  SSP Supported:	Yes
```

**Comment by avencherus** [comment]
@akien-mga 

Not getting vibration from PS4 controllers on Windows builds.  Checked and still present in 318c69351.

I have had this issue for as long as I have been using Godot.  Reluctant to report, since it seems like there are no longer active contributors for this area of the engine.

So the following line does nothing for me, but will work if I use an emulator to emulate an XBox controller.

`Input.start_joy_vibration(0, 1.0, 1.0, 1.0)`

**My details are:**

Win10 64-bit
Joypad - Sony DualShock 4
GUID - 4c05cc09000000000000504944564944

**Comment by follower** [comment]
Note that--on Windows--vibration relies on the [XInput API](https://docs.microsoft.com/en-us/windows/win32/xinput/getting-started-with-xinput) which the official docs describe as "allows applications to _receive input from the Xbox Controller_ for Windows"--i.e. at least officially, it's not a generic controller API.

I notice that [SDL haptic support](https://wiki.libsdl.org/CategoryForceFeedback) includes [XInput & DirectX Input support on Windows](http://hg.libsdl.org/SDL/file/78754c9488f6/src/haptic/windows), so adding additional controller support via DirectX may be possible.

Related issue: https://github.com/godotengine/godot/issues/30256

Related links:

 * My Steam Controller-specific research from researching an answer to this ["Steam controller vibration" query](https://old.reddit.com/r/godot/comments/f9jce9/steam_controller_vibration/fisohyu/).

    (Steam Controller controllers require specific haptic software (& firmware) support.)

 * Other related SDL source code links:

    * <http://hg.libsdl.org/SDL/file/78754c9488f6/src/haptic>

    * <http://hg.libsdl.org/SDL/file/78754c9488f6/src/joystick>

    * <http://hg.libsdl.org/SDL/file/78754c9488f6/src/joystick/steam>

**Comment by Lunatoid** [comment]
At least for Windows, controllers need to be emulated as an XInput device. I tested with my DualShock 4 controller on Windows 10 and it didn't work, however when I used DS4Windows, which turns your DS4 into an emulated Xbox controller, it did work.

DS4Windows seems to use the [ViGEmBus](https://github.com/ViGEm/ViGEmBus) driver for it's emulation, perhaps that could be a place to look.

**Comment by wingeisoftware** [comment]
For the record, I just ran into a "Vibration not working issue". My controller was plugged via a (cheap) usb hub, and that was the issue. I soon as i connected my gamepad directly to the computer it worked. I guess the hub did not deliver enough power for the vibration motors.

**Comment by djrain** [comment]
Vibration is not working for me using Xbox Series X controller, on macOS Big Sur over bluetooth.
The vibration _does_ work on this website https://gamepad-tester.com/ but not in Godot.

**Comment by trevorparscal** [comment]
It's also not working for my 8bitdo SM30 Pro+ (which emulates a Sony PS4 Dualshock) under macOS Monterey, and that same website works perfectly for me as well.

**Comment by umutseven92** [comment]
macOS Monterey 12.1, Macbook Pro M1 Max, Xbox One Controller over Bluetooth, Godot 3.4.2. `Input.start_joy_vibration` does not work.  Same as the commenter above, vibration works when tested via gamepad-tester.com.

**Comment by jeudy-ua** [comment]
For me, it works on a cheap USB controller on Ubuntu 21.10, but on Windows 10, same machine and controller, it doesn't work. Using version 3.3-stable.

**Comment by Stroopwafe1** [comment]
Same issue for me on Garuda Linux on kernal version 5.17.5-zen1-1-zen. Official controllers seem to work but my PDP Wired Controller for Xbox Series X - Black Camo doesn't seem to want to vibrate. It works fine on Windows so I imagine it's a drivers issue with my controller not being officially supported yet.
I added the SDL mapping to the input of our game, because the axes were mapped differently, but the vibration is done by something else I imagine...

**Comment by brettchalupa** [comment]
Is this issue up for grabs still? I'm interested in taking on trying to fix it. I've got a PS4 controller, Switch controller, and an Xbox One controller with access to Linux, macOS, and Windows. I'd be happy to dig into this and get it as cross-platform and cross-controller as possible.

**Comment by Rubonnek** [comment]
@brettchalupa It's up for grabs. I don't think anyone else is working on this.

**Comment by brettchalupa** [comment]
I've been doing some testing on Ubuntu, Steam Deck, and macOS today [with a little test app to detect controllers and rumble them](https://github.com/brettchalupa/grumble). Here are my findings so far with Godot v4.0.rc2.official [d2699dc7a].

- I couldn't get any controller to rumble on macOS 12.6. I tested the Xbox One, DualShock 4, and a Nintendo Switch Pro controllers.
- On Linux (Pop!_OS 22.04 LTS), the Xbox One controller rumbled. I couldn't connected my DualShock 4 or Nintendo Switch Pro controllers. I'd like to try those some more, as maybe Bluetooth was just being funky.
- The Xbox One controller rumbled on SteamOS. But surprisingly the Steam Deck's internal haptics didn't rumble. I have that functionality turned on and working in the config. Didn't work in desktop nor game modes of the OS.

I haven't booted up Windows and tested that yet, but I am assuming Xbox (aka Xinput) works there.

My next steps are:

- [ ] explore why controllers aren't rumbling on macOS and try to fix that
- [ ] further test out DualShock 4 controller on Linux and Windows, investigate findings
- [ ] look into why Steam Deck's haptics don't just work
- [ ] test rumble on iOS, Android, and web

If anyone else is interested in lending a hand with testing or investigating, let me know! It'd be much appreciated.

**Comment by HaSa1002** [comment]
On Windows, you'll find only XInput to be working. All other controllers need directinput that is not implemented. Switch Pro controllers only work using Steam Input when connected via cable. Bluetooth seems to be working though. I dunno what they do to make that work (Windows only reads information spam). SDL has some info how to trigger vibrations on DualShock and DualSense as well as an example how to use the adaptive triggers. Maybe you'll find some pointers there.

**Comment by brettchalupa** [comment]
@HaSa1002 Thanks, that's some helpful guidance and direction. 👍

**Comment by brettchalupa** [comment]
Let me know if I'm being too noisy here, but I thought I'd share some findings from the world of macOS rumble.

I tested some Unity games that I know have rumble support and none of them worked on macOS. So I dug into Apple's haptics API and can confirm it rumbles DS4 and XInput controllers [with this sample game they provided](https://developer.apple.com/library/archive/samplecode/scenekit-2017/Introduction/Intro.html#//apple_ref/doc/uid/TP40017656). So at the very baseline, we know that with Swift and Apple's SDK, you can rumble a DS4 and Xbox One controller on macOS 12.6.

Then I thought, okay, let's see if SDL2 can rumble controllers on macOS. [I wrote a little SDL2 CLI tester](https://github.com/brettchalupa/sdl_rumble) and tested some controllers. The controllers using XInput rumbled, awesome! And a controller that emulates Switch Pro Controller rumbled too, just a bit weakly. DS4/MFi and DirectInput controllers didn't rumble though, and SDL reports them as not supporting rumble.

So what does this mean? Well, it should at least be possible to get XInput controllers rumbling on macOS in Godot by referencing SDL's implementation. That's my initial goal, that way XInput controllers work on macOS, Windows, and Linux.

From there, outstanding questions are:

- How do we get DS4 (and maybe Dual Sense?) controllers working with rumble triggers too? This seems like an interesting resource: https://github.com/JibbSmart/JoyShockLibrary
- Test on mobile
- Further explore different controller APIs and rumble support for all of them on as many platforms as possible (Switch Pro Controllers, DirectInput, etc.)

But I think XInput on modern macOS probably seems like a fine starting place for now for making this a bit better.

All right, time to dig into the Godot source. 😸

**Comment by Calinou** [comment]
> How do we get DS4 (and maybe Dual Sense?) controllers working with rumble (and maybe the triggers too)? This seems like an interesting resource: [JibbSmart/JoyShockLibrary](https://github.com/JibbSmart/JoyShockLibrary)

See discussion on https://github.com/godotengine/godot-proposals/issues/2829. JoyShockLibrary is in maintenance mode and may not remain maintained in the long term, but we don't want to adopt the entirety of SDL for binary size reasons (among other concerns). Copy-pasting code from relevant libraries is fine still.

**Comment by brettchalupa** [comment]
I’ve continued my research and testing to better understand what’s possible and what is and isn’t working. Here are some more findings:

- Windows:
   - XInput rumbles without issue in Godot on Windows. 
   - DualShock 4 does not, as others have reported. [SDL2 does with a special hint.](https://gist.github.com/brettchalupa/758a6f98e2b6285dec8fa6568715ce67#file-main-cpp-L9) So it should be possible to add to Godot. DS4Windows and Steam Input seem like potentially viable alternatives for immediate needs. I tested Steam Input with DS4 and confirmed rumbles work in a Godot game. A lot of people have DS4 controllers so I want to dig into this working on Windows without a need for Steam or DS4Windows. 
- MacOS:
  - Steam Input has no effect on rumble on XInput or DS4 in a Godot game running through it from my testing. But Steam can rumble everything when identifying controllers. Steam certainly has some advanced rumble integration from what I can tell. I wonder if any of it is open source…
- Linux:
  - Did more testing and XInput, Switch, and DualShock 4 are all rumbling for me without issue on Ubuntu 22.04. Yay. 
  - My Steam Deck appears to be having issues with all haptics and rumble, so I am going to try to fix that and then will do more Godot rumble testing on the Deck. Connected DS4, XInput, and Switch controllers rumble fine. 
- Web
  - Only Chrome (and thus Edge) supports rumble from testing on Mac, Linux, and Windows vis the JS console. [The API is pretty simple.](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/vibrationActuator)
  - Godot web exports don’t rumble in Chrome. Seems possible though. It’s not a stable, standard API, so maybe risky to add? Would like to experiment with this more. 
  
In summary: rumble for XInput on MacOS and DualShock 4 on Windows are my highest priorities.

Need to test Deck, iOS, and Android more. Chrome web may be possible. I don’t have a Dual Sense controller yet, but I’d like to eventually test and support that too.

A lot of what I’ve shared is just confirming the issues reported above, but it helps me to test the issues and have a couple control groups to test against like the SDL tester and a Unity game I know rumbles. I have a pretty good grasp on what’s possible and what to do next. Will try to report back in the coming days and hopefully have a PR.

**Comment by brettchalupa** [comment]
I've been digging into the Godot MacOS joypad vibration code and have some options for the path forward. Feedback and guidance would be much appreciated!

## Current Godot MacOS Joypad Vibration Implementation

Right now the Godot Joypad implementation on MacOS uses Apple's [ForceFeedback](https://developer.apple.com/documentation/forcefeedback?language=objc) API. See [the headers](https://github.com/godotengine/godot/blob/6f64349bfe42b795a6c50e9c2093d38e7a992e7a/platform/macos/joypad_macos.h#L34-L35) and [vibration code](https://github.com/godotengine/godot/blob/6f64349bfe42b795a6c50e9c2093d38e7a992e7a/platform/macos/joypad_macos.cpp#L505-L506). The ForceFeedback API has been supported in MacOS since 10.2, released back in 2003. It's not clear to me what devices this rumbles, but we do know it doesn't work for modern game controllers.

Godot also [uses the MacOS IOKit API](https://github.com/godotengine/godot/blob/6f64349bfe42b795a6c50e9c2093d38e7a992e7a/platform/macos/joypad_macos.h#L36) for interfacing with joypads.

## MacOS Controller Paths Forward

Here are the options I see:

1. **[Use the MacOS GameController system framework](https://developer.apple.com/documentation/gamecontroller?language=objc)** — this is Apple's robust controller API for interacting with modern controllers in all sorts of ways, including vibration/haptics. This is what [Godot's iOS Joypad implementation uses](https://github.com/godotengine/godot/blob/6f64349bfe42b795a6c50e9c2093d38e7a992e7a/platform/ios/joypad_ios.h#L31).
    - Compatibility: iOS 7.0+, iPadOS 7.0+, macOS 10.9+, Mac Catalyst 13.0+, tvOS 9.0+ 
    - Would be implemented in Obj-C++, like iOS Joypad
    - Pros:
      - Will work with Xbox One, PS4, etc. controllers.
      - System API
      - Similar and/or shared implementation with Godot's iOS Joypad
      - It will be straightforward to get vibration working on iOS as a follow up
    - Cons:
      - Would require a re-implementation of the current MacOS joypad implementation, can't just add in a rumble support.
      - Might mean dropping support for MacOS joypad support for prior operating systems prior to macOS 10.9 (2013 release), unless there's a desire to maintain the current joypad implementation for those older operating system versions.
2. **[Use the HIDAPI library](https://github.com/libusb/hidapi)** — HIDAPI is a liberally [licensed](https://github.com/libusb/hidapi/blob/master/LICENSE.txt) Human Interface Device (HID) library for interfacing with a variety of controllers. [This is used by SDL](https://github.com/libsdl-org/SDL/blob/release-2.26.3/src/hidapi/mac/hid.c), and it seems to work quite well for wide compatibility.
    - Pros: 
      - Mature, stable library that's been widely used
      - Could help enable more than XInput rumble on Windows?
    - Cons:
      - Pulling in an entire library. Would Godot just use it for MacOS or for everything?
      - Requires a re-implementation without any reference within Godot

## My Thoughts

I am honestly not sure... HIDAPI seems like it could unlock DualShock 4 rumble on Windows but would require a lot of work. More than I am likely capable of taking on. Going with the Apple GameController framework seem approachable because of the iOS joypad implementation I can reference. It's also appealing to use a system API that can be shared across Apple platforms.

## Next Steps

I'm not an expert in this domain. I'm just a person who wants my controller to rumble and am willing to do the work to help make that happen. 😂 I may be missing an option for a path forward.

Are any contributors/maintainers able to give me some guidance on how to proceed with this work or conversation? The scope of the decision seems too big for me to just make and start hacking away at the implementation without guidance/feedback.

**Comment by Calinou** [comment]
> Might mean dropping support for MacOS joypad support for prior operating systems prior to macOS 10.9 (2013 release), unless there's a desire to maintain the current joypad implementation for those older operating system versions.

Godot 4.0 requires macOS 10.12 or later, so that's fine.

**Comment by brettchalupa** [comment]
Given I haven't gotten much feedback on my comment above, I'll plan on proceeding with the proposed option 1, **Use the MacOS GameController system framework** for getting MacOS vibration working. After sitting with this and learning more Godot over the past month, I think it's a good long-term solution.

Once that's done, I'd like to explore vibration on iOS, Android, Chrome web, and DualShock 4 on Windows after that. Help/support with DS4 on Windows would be welcome, as it's quite a different implementation than what exists.

Just to set expectations, it may be a few weeks/months before I have time to do this. But it's on my radar and important to me as I develop primarily on MacOS and want to add and test rumble in my games.

**Comment by 27thLiz** [comment]
that should be fine, provided you can keep compatibility with the existing gamepad mappings

**Comment by brettchalupa** [comment]
Sorry for such silence here on this issue. I sadly no longer have a macOS machine due to a job change and am not able to really dig into the issue of getting rumble working on modern macOS. I hope that my findings that I shared above at least help someone pick this up with a little bit of context and info. Consider this issue unassigned and free to pick up for anyone who is interested in attempting to fix it. :+1:

**Comment by JezerM** [comment]
Hi, I just noticed that vibration doesn't work for controllers in macOS a few days ago, so I tried to fix it. I would appreciate if someone interested could test it.

And thanks to @brettchalupa for the initial guide on how controllers work in Godot and how this issue should be addressed.

**Comment by coelhucas** [comment]
Tested on MacOS with Xbox Series S gamepad, dualshock 4 and joycons, none working (as of version 4.1.3).

**Comment by Alex2782** [comment]
@coelhucas There is already PR, I was able to test my Bluetooth XBox controller on Macos.
https://github.com/godotengine/godot/pull/80709

**Comment by JezerM** [comment]
For macOS, this is already fixed in the master branch~

**Comment by akien-mga** [comment]
After https://github.com/godotengine/godot/pull/80709, I think we can close this old issue.

If anyone still experiences non-working vibration with some controllers on some platforms in 4.3 or later, please open a dedicated issue for that controller / platform combination so it can be assessed.

**Comment by Yurinka** [comment]
The issue continues happening in v4.2.2.stable.official when running it (reproduced it in the editor, using Windows 10):

-Xbox compatible gamepads rumble properly
-DualShock4 (connected via USB) doesn't rumble
-DualSense (connected via USB) doesn't rumble
-Switch Pro controller doesn't rumble

**Comment by timenoe** [comment]
I am having the same issue on 4.3.stable on Windows 11. I've tried multiple Sony/Nintendo controllers and none seem to provide vibration. All of my Xbox gamepads work fine. Seems to be a DirectInput vs. XInput sort of thing.

**Comment by Sjoerd1988** [comment]
Godot 4.4 stable, rumble still does not work for PS5 controller.
